content/src/logic/snapshot-cleaner.ts

```diff
@@ -0,0 +1,95 @@
+import PQueue from 'p-queue'
+import path from 'path'
+import { promisify } from 'util'
+import { streamToBuffer } from '../ports/contentStorage/contentStorage'
+import { FSComponent } from '../ports/fs'
+import { AppComponents } from '../types'
+import { filterContentFileHashes } from './database-queries/snapshot-cleaner-queries'
+
+export async function cleanSnapshots(
+  components: Pick<AppComponents, 'logs' | 'gzipCompressor' | 'database'> & {
+    fs: Pick<FSComponent, 'unlink' | 'createReadStream'>
+  },
+  storageDirectory: string,
+  minimumSnapshotSizeInBytes: number
+): Promise<void> {
+  return cleanSnapshotsWithExec(components, promisify(exec), storageDirectory, minimumSnapshotSizeInBytes)
+}
+
+// This function that takes exectueCommand that is for injecting a mock and being able to unit test.
+// Didn't found a way to mock exec/promisify functions.
+export async function cleanSnapshotsWithExec(
+  components: Pick<AppComponents, 'logs' | 'gzipCompressor' | 'database'> & {
+    fs: Pick<FSComponent, 'unlink' | 'createReadStream'>
+  },
+  executeCommand: (command: string) => Promise<{ stdout: string; stderr: string }>,
+  storageDirectory: string,
+  minimumSnapshotSizeInBytes: number
+): Promise<void> {
+  const logger = components.logs.getLogger('Snapshot Cleaner')
+
+  const isModernSnapshotHeader = (header: string): boolean => header.startsWith('### Decentraland json snapshot')
+  const isLegacySnapshotHeader = (header: string): boolean => header.match(/\[\[\"Qm.+?(?=\",\[\")/) !== null
+
+  const isSnapshot = async (filepath: string): Promise<boolean> => {
+    let readStream: ReadStream | undefined
+    try {
+      readStream = components.fs.createReadStream(filepath, { end: 59 })
+      const header = await streamToBuffer(readStream)
+      const potentialSnapshotHeader = header.toString('utf8')
+      return isModernSnapshotHeader(potentialSnapshotHeader) || isLegacySnapshotHeader(potentialSnapshotHeader)
+    } finally {
+      if (readStream) {
+        readStream.close()
+      }
+    }
+  }
+
+  const deleteIfItIsASnapshot = async (filepath: string): Promise<void> => {
+    try {
+      const isCompressed = filepath.endsWith('.gzip')
+      const potentialSnapshotFilepath = isCompressed ? filepath.slice(0, -5) : filepath
+      if (isCompressed) {
+        await components.gzipCompressor.decompress(filepath, potentialSnapshotFilepath)
+      }
+      if (await isSnapshot(potentialSnapshotFilepath)) {
+        logger.debug(`Deleting snapshot: ${filepath}`)
+        await components.fs.unlink(filepath)
+      }
+      if (isCompressed) {
+        await components.fs.unlink(potentialSnapshotFilepath)
+      }
+    } catch (error) {
+      logger.error(error, { filepath })
+    }
+  }
+
+  const getPathsOfBigFiles = async (directory: string): Promise<string[]> => {
+    const { stdout, stderr } = await executeCommand(
+      `find ${directory} -type f -size +${minimumSnapshotSizeInBytes - 1}c`
```

Comment:
This doesn't work in windows

Reply:
Won't work even if the docker container is running linux?

Reply:
Regarding windows, we had issues created by folks running catalysts in windows. I don't think we should close that door.

Mmm I can think about some alternatives, but not a clear path ahead...

- how long does it take to execute that `find ...` command in the catalyst?
- how long would `glob('*/*')` take? (maybe it may be part of the storage component, for consistency and abstraction)

those answers may lead to different implementation questions like:

- how long would take to run `storage.allEntries().filter(db.isActiveFile)`, skipping the stat+uncompress?
- how long `let bloom = bloomFilter(db.activeFiles); storage.allEntries().filter(not in bloom).filter(db.isActiveFile).map(storage.deleteFile)` would take?

Reply:
This script is a one-time run, as we need to delete some forgotten snapshots files, but with the bump of the snapshots-fetcher version, we won't get any more of them. So, as long as this code doesn't crash on Windows we are good to go. Maybe add a try/catch?